File: scope_policy.py
def missing_scope_values(granted_scopes, required_scopes):
    """付与済みスコープから不足スコープ値の一覧を返す。"""
    granted = {str(scope) for scope in (granted_scopes or [])}
    missing = []
    for scope in required_scopes:
        value = getattr(scope, "value", str(scope))
        if value not in granted:
            missing.append(value)
    return missing

File: src/auth/scope-policy.ts
/**
 * OAuth スコープポリシー
 */

/**
 * 付与済みスコープから不足しているスコープ値を返す
 */
export function missingScopeValues(
  grantedScopes: string[],
  requiredScopes: string[]
): string[] {
  const grantedSet = new Set(grantedScopes);
  return requiredScopes.filter((scope) => !grantedSet.has(scope));
}

/**
 * 付与済みスコープから有効なスコープリストを返す
 * 空の場合はデフォルトを返す
 */
export function activeAuthScopesFromGranted(
  grantedScopes: string[],
  defaultScopes: string[]
): string[] {
  if (grantedScopes.length === 0) {
    return [...defaultScopes];
  }
  return [...new Set(grantedScopes)];
}

/**
 * スコープ値をソートして正規化する
 */
export function normalizeScopeValues(scopes: string[]): string[] {
  return [...new Set(scopes)].sort();
}
